crates/polars-plan/src/dsl/function_expr/temporal.rs

```diff
@@ -189,33 +189,53 @@ pub(super) fn datetime(
 fn apply_offsets_to_datetime(
     datetime: &Logical<DatetimeType, Int64Type>,
     offsets: &StringChunked,
-    offset_fn: fn(&Duration, i64, Option<&Tz>) -> PolarsResult<i64>,
     time_zone: Option<&Tz>,
 ) -> PolarsResult<Int64Chunked> {
-    match (datetime.len(), offsets.len()) {
-        (1, _) => match datetime.0.get(0) {
-            Some(dt) => offsets.try_apply_nonnull_values_generic(|offset| {
-                offset_fn(&Duration::parse(offset), dt, time_zone)
-            }),
-            _ => Ok(Int64Chunked::full_null(datetime.0.name(), offsets.len())),
-        },
-        (_, 1) => match offsets.get(0) {
+    match offsets.len() {
+        1 => match offsets.get(0) {
             Some(offset) => {
                 let offset = &Duration::parse(offset);
-                datetime
-                    .0
-                    .try_apply_nonnull_values_generic(|v| offset_fn(offset, v, time_zone))
+                if offset.is_constant_duration(datetime.time_zone().as_deref()) {
+                    // fastpath!
+                    let duration = match datetime.time_unit() {
+                        TimeUnit::Milliseconds => offset.duration_ms(),
+                        TimeUnit::Microseconds => offset.duration_us(),
+                        TimeUnit::Nanoseconds => offset.duration_ns(),
+                    };
+                    if offset.negative() {
```

Comment:
To reduce binary bloat, I think we can make one code branch of this that does the `apply_values`.

```rust
if offset.negative() {
   duration = - duration;
}

Ok(datetime.0.apply_values(|v| v + duration))
```